Alignment/APEEstimation/plugins/createDefaultApeTree.cc

```diff
@@ -0,0 +1,474 @@
+{
+    
+  if(v_id.empty())return true;
+  for(std::vector<unsigned int>::const_iterator i_id = v_id.begin(); i_id != v_id.end(); ++i_id){
+    if(id==*i_id)return true;
+  }
+  return false;
+}
+
+bool
+DefaultApeTree::checkModuleBools(const bool id, const std::vector<unsigned int>& v_id)const
+{
+    
+  if(v_id.empty())return true;
+  for(std::vector<unsigned int>::const_iterator i_id = v_id.begin(); i_id != v_id.end(); ++i_id){
+    if(1==*i_id && id)return true;
+    if(2==*i_id && !id)return true;
+  }
+  return false;
+}
+
+bool
+DefaultApeTree::checkModuleDirections(const int id, const std::vector<int>& v_id)const
+{
+    
+  if(v_id.empty())return true;
+  for(std::vector<int>::const_iterator i_id = v_id.begin(); i_id != v_id.end(); ++i_id){
+    if(id==*i_id)return true;
+  }
+  return false;
+}
+
+bool
+DefaultApeTree::checkModulePositions(const float id, const std::vector<double>& v_id)const
+{
+    
+  if(v_id.empty())return true;
+  int entry(1); double intervalBegin(999.);
+  for(std::vector<double>::const_iterator i_id = v_id.begin(); i_id != v_id.end(); ++i_id, ++entry){
+    if(entry%2==1)intervalBegin = *i_id;
+    if(entry%2==0 && id>=intervalBegin && id<*i_id)return true;
+  }
+  return false;
+}
+
+
+
+// ------------ method called to for each event  ------------
+void
+DefaultApeTree::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  // Same procedure as in ApeEstimatorSummary.cc minus reading of baseline tree
+  
+  // Load APEs from the GT and write them to root files similar to the ones from calculateAPE()
+  edm::ESHandle<AlignmentErrorsExtended> alignmentErrors;
+  iSetup.get<TrackerAlignmentErrorExtendedRcd>().get(alignmentErrors);  
+  
+  // Set up root file for default APE values
+  const std::string defaultFileName(parameterSet_.getParameter<std::string>("resultFile"));
+  TFile* defaultFile = new TFile(defaultFileName.c_str(),"RECREATE");
+  
+  // Naming in the root files has to be iterTreeX to be consistent for the plotting tool
+  TTree* defaultTreeX(nullptr);
+  TTree* defaultTreeY(nullptr);
+  defaultFile->GetObject("iterTreeX;1",defaultTreeX);
+  defaultFile->GetObject("iterTreeY;1",defaultTreeY);
+  // The same for TTree containing the names of the sectors (no additional check, since always handled exactly as defaultTree)
+  TTree* sectorNameTree(nullptr);
+  defaultFile->GetObject("nameTree;1",sectorNameTree);
+  
+  edm::LogInfo("DefaultAPETree")<<"APE Tree is being created";
+  defaultTreeX = new TTree("iterTreeX","Tree for default APE x values from GT");
+  defaultTreeY = new TTree("iterTreeY","Tree for default APE y values from GT");     
+  sectorNameTree = new TTree("nameTree","Tree with names of sectors");
+  
+    
+  // Assign the information stored in the trees to arrays
+  double a_defaultSectorX[16589];
+  double a_defaultSectorY[16589];
+  
+  std::string* a_sectorName[16589];
```

Comment:
@mteroerd Why 16589? I think this is the number of modules in the phase-0 tracker, i.e. it was kind of an upper limit on the sector size, but now it isn't anymore. In general you could use a std::vector, but this likely needs to be propagated to the other APE tools, too.

Reply:
Since I build the sectors in this plugin, I think I can just take the number of sectors from there instead of using such a large number by default. However, it seems like just using vectors is not that easy when defining the branches:
defaultTreeX->Branch(ss_sector.str().c_str(), &a_defaultSectorX[iSector], ss_sectorSuffixed.str().c_str());
It seems like you cannot use vectors the way that arrays are used here because when filling the TTree, a segmentation fault is thrown. However, maybe I am missing something.

Reply:
@mteroerd please have a look at [this example](https://root.cern.ch/root/html/tutorials/tree/hvector.C.html). Bottom line: you have to use a pointer to the vector.
For a string it is similar and you can find a possible implementation including reading and writing to the tree [here](https://github.com/cms-sw/cmssw/commit/ddfa7075ab48927d085a4ba5bb720dadfcf1ef6d).

Hope this helps.

Reply:
This example does not really work here. It is for writing a vector to a tree. In my case, I only write different elements from an array to different branches. So what has to be done with the vector is write each element of the vector to a separate branch, so instead of storing a_defaultSectorX, we are storing a_defaultSectorX[iSector]. But for vectors, &a_defaultSectorX[iSector-1] does not seem to work when Using TTree->Branch().

Reply:
hm, ok, then maybe you should dynamically allocate (and later delete) the array
this way you can use a dynamic size of the array
but maybe, people more expert on ROOT's behaviour can comment here...
Randomly picking people you might have ideas: @pcanal, @Dr15Jones

Reply:
As far as I know, dynamic arrays are allocated with a fixed size, too. So there is no reason to dynamically allocate the arrays, since if I know the size, I can just use a regular array with that size.